🚀 1단계 - 스프링 프레임워크 적용

hongmoSung opened the pull request:
## 요구사항

- [x] 웹 요청/응답 구현하기
- [x] DB 연동하기

hongmoSung committed:
feat: 🚀 1단계 - 웹 요청/응답 구현하기

hongmoSung committed:
feat: 🚀 1단계 - DB 연동하기

qhals321 commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,56 @@
+                maxPosition = racingCar.getPosition();
+                winners.clear();
+            }
+            if (racingCar.getPosition() >= maxPosition) {
```
maxPosition 보다 클 수 있는 상황이 있나요?? 🤔🤔🤔

hongmoSung replied:
클 수 있는 상황이 없을거 같네요... legacy sample 코드를 가져오면서 확인을 하지 않았습니다.. 😅
가독성을 위해 stream 으로 변경하면서 수정하였습니다.

qhals321 commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,56 @@
+                .collect(Collectors.toList());
+        int count = gameStartDto.getCount();
+
+        IntStream.rangeClosed(1, count).forEach((i) -> playRound(racingCars));
```
Stream 활용이 어마어마하시네요! 💯

hongmoSung replied:
감사합니다 👍

qhals321 commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,56 @@
+
+        int maxPosition = 0;
+        List<String> winners = new ArrayList<>();
+        for (RacingCar racingCar : racingCars) {
+            if (racingCar.getPosition() > maxPosition) {
+                maxPosition = racingCar.getPosition();
+                winners.clear();
+            }
+            if (racingCar.getPosition() >= maxPosition) {
+                winners.add(racingCar.getName());
+            }
+        }
```
성능 면에서 굉장히 좋은 코드네요!! 💯💯💯

아니면 racingCar 가 정말 아주아주아주 사이즈가 큰 부분이 아니라면 Stream 을 이용해 가독성을 챙겨줘도 좋을 것 같아요!! (해당 부분은 항상 선택인 것 같네요 🥹)

성능 부분에서는 지금 단계에서 큰 차이가 없을 것 같으니 한 번 Stream 으로 가독성을 챙겨보는 건 어떨까요?!

qhals321 reviewed (changes requested):
안녕하세요 홍모님! 스프링 프레임워크 적용 정말 잘해주셨네요!! 💯💯💯

스프링 활용과 자바 활용이 엄청나네요! (특히 스트림 활용이 어마어마하시네요!) 👍👍

홍모님만의 레이어 역할 정의가 궁금해요!! 🤔 관련해서는 코드리뷰는 남기지 않았어요! 이번 단계에서는 역할 정의만 해주시고 다음 단계에서 확실하게 레이어를 나눠봐요!! (만일 해당 역할 정의에 어울리지 않는 코드가 있다면 리팩토링 도전도 환영이에요 🥹) 

또한, 현재 서비스는 테스트가 불가능한 상태로 보이는데 테스트를 작성해보시면서 테스트를 작성 가능한 서비스로 만들어보는 것도 좋을 것 같아요!! 🙇🙇🙇

**정리**
- 스프링, 자바 활용 최고 👍👍
- 각 레이어별 테스트 코드 작성하기
- 레이어 역할 정의하기

hongmoSung committed:
fix: maxPosition 수정

- https://github.com/next-step/spring-racingcar/pull/51#discussion_r1181155132

hongmoSung committed:
refactor: layer 나누기

- https://github.com/next-step/spring-racingcar/pull/51#pullrequestreview-1407045565

hongmoSung committed:
refactor: test 코드 작성을 위해 리펙토링

- https://github.com/next-step/spring-racingcar/pull/51#discussion_r1181155838 가독성을 위한 리펙토링
- 테스트 코드 작성을 위해 비즈니스 로직 메소드로 추출
- 테스트 코드 작성을 위해 생성자 작성

hongmoSung committed:
test: 비즈니스 레이어 테스트 코드 작성

hongmoSung committed:
test: 데이터 레이어 테스트 코드 작성

hongmoSung committed:
test: 프레센테이션 레이어 테스트 코드 작성

qhals321 commented on src/test/java/racingcar/business/RacingServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+        this.racingService = new RacingService(racingRepository);
+    }
+
+    @DisplayName("자동차 이름 입력값으로 반환된 자동차 갯수를 확인한다.")
```
DisplayName 으로 어떤 테스트를 하는지 확실하게 표현해주니 너무 좋네요~! 👍

qhals321 commented on src/test/java/racingcar/business/RacingServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+    @DisplayName("자동차 이름 입력값으로 반환된 자동차 갯수를 확인한다.")
+    @Test
+    void getRacingCarsTest() {
+        // given
+        String names = "A,B,C,D,E";
+        GameStartDto gameStartDto = new GameStartDto(names, 0);
+
+        // when
+        List<RacingCar> racingCars = racingService.getRacingCars(gameStartDto);
+
+        // then
+        int playerCount = names.split(",").length;
+        assertThat(racingCars.size()).isEqualTo(playerCount);
```
given, when, then 테스트까지~! 👍

qhals321 commented on build.gradle:
```diff
@@ -12,7 +12,15 @@ repositories {
+	runtimeOnly 'com.h2database:h2'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+
+	testImplementation 'junit:junit:4.13.1'
+	testRuntimeOnly 'junit:junit:4.13.1'
+
+	testImplementation 'org.assertj:assertj-core:3.23.1'
+	testRuntimeOnly 'org.assertj:assertj-core:3.23.1'
```
`spring-boot-starter-test` 의존성 이미 가져와서 해당 의존성들은 필요 없지 않나용??

qhals321 commented on src/test/java/racingcar/data/RacingRepositoryTest.java:
```diff
@@ -0,0 +1,50 @@
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrowsExactly;
+
+@ExtendWith(SpringExtension.class)
```
해당 어노테이션은 지워도 될 것 같아요!!
`@SpringBootTest` 내부를 살펴보면 해당 어노테이션이 그대로 선언되어 있는 걸 볼 수 있을 거에요!
<img width="549" alt="image" src="https://user-images.githubusercontent.com/63535027/235358619-4274b6e0-61d0-4fb5-9927-9ec61f85d096.png">

qhals321 commented on src/test/java/racingcar/business/RacingServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+class RacingServiceTest {
+
+    private final RacingService racingService;
+
+    public RacingServiceTest() {
+        JdbcTemplate jdbcTemplate = new JdbcTemplate();
+        RacingRepository racingRepository = new RacingRepository(jdbcTemplate);
+        this.racingService = new RacingService(racingRepository);
+    }
```
서비스 테스트도 repositoryTest 처럼 `@SpringBootTest` 를 이용해 통합테스트를 만들어도 괜찮을 것 같아요!!

또한, 레이어를 독립적으로 테스트를 하고 싶으시다면 [슬라이스 테스트](https://tecoble.techcourse.co.kr/post/2021-05-18-slice-test/) 를 한 번 살펴봐도 좋을 것 같아요!

(두 방법 중 본인의 취향에 맞게 테스트를 작성해나가면 좋을 것 같아요!)

qhals321 commented on src/main/java/racingcar/business/RacingService.java:
```diff
@@ -0,0 +1,61 @@
+        return gameResultDto;
+    }
+
+    public int startCarRacingGame(List<RacingCar> racingCars, int count) {
+        return IntStream.rangeClosed(1, count).reduce(0, (acc, n) -> {
+            this.playRound(racingCars);
+            return acc + 1;
+        });
+    }
+
+    public List<RacingCar> getRacingCars(GameStartDto gameStartDto) {
+        return Stream.of(gameStartDto.getNames().split(",")).map(RacingCar::new)
+                .collect(Collectors.toList());
+    }
```
혹시 테스트 때문에 public 으로 열어둔 걸까요?! 테스트 때문에 접근제어를 열어두는 것보다 객체의 책임을 나누어 해당 기능들을 나눠주면 어떨까요?!

예를 들어, `List<RacingCar>` 를 담고 있는 `RacingCars` 라는 [일급컬렉션](https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/) 을 만들어 생성자에 `names` 가 들어가면 split 을 해줘서 필드에 매핑해줘서 해당 생성자 테스트를 만들면 조금 더 역할 분리가 될 것 같아요!

qhals321 commented on src/test/java/racingcar/business/RacingServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+class RacingServiceTest {
```
제가 이전에 리뷰를 남겼던 부분은 `game` 기능에 대한 테스트를 작성하는 부분이었어요! 결국 해당 기능에서 테스트가 되어야 하는 부분은 "가장 많이 간 player 가 우승한다." 라는 부분인데 아무래도 해당 기능 안에 "랜덤 숫자 생성" 이라는 랜덤성 때문에 테스트가 힘드셨을 거에요! 

'랜덤 생성기'와 '정해진 숫자 생성기' 를 원할 때마다 `game` 이라는 메서드가 선택해서 사용할 수 있는 구조로 만들 수 있는 방법을 고민해보시는 것도 좋을 것 같아요!! (여기서 키워드는 '다형성' 혹은 '[전략패턴](https://tecoble.techcourse.co.kr/post/2021-10-04-strategy-command-pattern/)' 입니다!)

만일 위와 같은 구조가 되었다면 테스트가 가능해질 거에요!

고민해보시고 언제든 궁금한 부분이 생기신다면 dm 도 환영합니다~! 🙇🙇🙇

qhals321 commented on src/test/java/racingcar/data/RacingRepositoryTest.java:
```diff
@@ -0,0 +1,50 @@
+        GameResultDto gameResultDto = new GameResultDto(winners, racingCars);
+
+        // when
+        assertThrowsExactly(DataIntegrityViolationException.class,
```
예외 테스트까지 👍👍👍👍

qhals321 reviewed (approved):
빠른 리뷰 적용~! 최고네요! 👍👍👍👍

테스트를 처음 작성해보신다고 하셨는데 예외케이스까지 챙기시다니.. 최고네요 💯

몇 가지 코멘트 남겼습니다! (관련 키워드와 링크도 추가해뒀어요~!) 2단계 진행하시면서 시간 나실 때 적용해보는 것도 좋을 것 같아요!! 🔥🔥🔥 코멘트가 이해가 되지 않으시거나 궁금한 부분이 생기신다면 언제든 dm 주세요!! 환영합니다 🫡

레이어 관련해서는 다음 단계에서 더 자세히 다뤄봐요!! 🙇 1단계 수고 많으셨습니다~! 🙇🙇🙇
